src/smt_util/node_traversal.cpp

```diff
@@ -0,0 +1,72 @@
+#include "node_traversal.h"
+
+namespace CVC4 {
+
+NodePostorderIterator::NodePostorderIterator(TNode n) : d_stack{{false, n}}
+{
+  enqueueChildrenUntilLeaf();
+}
+
+NodePostorderIterator::NodePostorderIterator() : d_stack() {}
+
+NodePostorderIterator& NodePostorderIterator::operator++()
+{
+  if (d_stack.size() > 0)
+  {
+    d_stack.pop_back();
+    enqueueChildrenUntilLeaf();
+  }
+  return *this;
+}
+
+NodePostorderIterator NodePostorderIterator::operator++(int)
+{
+  NodePostorderIterator copyOfOld(*this);
+  ++*this;
+  return copyOfOld;
+}
+
+TNode& NodePostorderIterator::operator*() { return d_stack.back().second; }
+
+bool NodePostorderIterator::operator==(const NodePostorderIterator& other) const
+{
+  return d_stack == other.d_stack;
+}
+
+bool NodePostorderIterator::operator!=(const NodePostorderIterator& other) const
+{
+  return !(*this == other);
+}
+
+void NodePostorderIterator::enqueueChildrenUntilLeaf()
+{
+  if (d_stack.size() > 0)
+  {
+    // While the top element does not have its children enqueued,
+    while (!d_stack.back().first)
+    {
+      d_stack.back().first = true;
+      TNode back = d_stack.back().second;
+      // use integer underflow wrapping behavior to reverse-iterate over the
+      // children,
+      for (size_t i = back.getNumChildren() - 1; i < back.getNumChildren(); --i)
+      {
+        // and enqueue them.
+        d_stack.emplace_back(false, back[i]);
+      }
+    }
```

Comment:
This looks like you are doing a tree traversal here, since you don't prevent nodes from being pushed multiple times onto `d_stack`. You need an additional cache for this that keeps track of visited nodes (which should make the bool flag on the stack obsolete). You usually don't want to do tree traversals on DAGs.

Reply:
Thanks! You're completely right. I forgot that nodes are DAGs. Fixed in 5b4bcc498.

I believe that just a visit set is insufficient, and I used a map instead, but I'd love to find a way to do without the map.